traversal/planner/PlannerEdge.java

```diff
@@ -231,7 +247,14 @@ private void recordCost() {
 
         protected void updateOptimiserCoefficients() {
             assert costLastRecorded == safeCost();
-            planner.optimiser().setObjectiveCoefficient(varIsMinimal, log(1 + safeCost()));
+            
+            double[] filters = to.ins().stream().filter(e -> !e.equals(this)).filter(this::cheaperThan).mapToDouble(e -> e.safeCost() / to.safeCost()).toArray();
+            double averageFilter = 1.0;
+            if (filters.length > 0) averageFilter = pow(Arrays.stream(filters).reduce(1.0, (a,b) -> a*b), 1.0 / filters.length);
```

Comment:
Geometric mean of the "filtering power" of all the adjacent edges which are more expensive than this one. The filtering power is defined as the edge's cardinality divided by the cardinality of its destination vertex, i.e. "what portion of all valid concepts will traversing this edge match?".